Implement modular arithmetic

jellevos opened the pull request:
Hi @tarcieri and the other contributors! For the `scicrypt` crate I was working on constant-time big integers by wrapping GMP, but the interface was extremely clumsy. I decided to switch to this crate (which looks awesome!) and added the necessary methods for modular arithmetic.

I implemented the following functionality:
- Shifting left and right by 1 and returning the overflow as a `Choice`
- `Neg` for `Wrapping<Uint<LIMBS>>`
- Modular multiplicative inverse
- Computing parameters for the Montgomery form (MF)
- Going to and from MF
- Modular addition (in MF)
- Modular multiplication (in MF)
- Modular exponentiation (in MF)

In my opinion, there are at least two large issues right now that I would like to hear your opinion on.
- [x] The `Modular` interface is not satisfying to me because it copies the `MontgomeryParameters` and does not assert that two `Modular` integers indeed share the same modulus.
- [x] None of the functions are currently `const fn`.

Still, I thought it was good to already make a PR to bring this to your attention. I am happy to put in the effort to bring this to a point where it is mergeable! :)

**Edit:**
The modulus is now a constant and the struct is called `Residue` rather than `Modular`.

jellevos commented:
I now see issues #70 and #28. I think `adt_const_params` will take quite a while before it is stabilized, but in the meantime one option is not to use a constant modulus and instead use debug asserts to check that e.g. the two moduli in a multiplication are the same. Based on these two issues I will change `Modular` to `Residue` and change as many functions as possible to be `const`.

tarcieri commented:
@jellevos the workaround that @haslersn suggested which uses a ZST + trait with an associated constant seems OK as a workaround until `adt_const_params` is stabilized:

https://github.com/RustCrypto/crypto-bigint/issues/70#issuecomment-1193319917

```rust
pub trait ResidueParameters<const L: usize>: PartialEq + Debug {
    const MODULUS: Option<UInt<L>>;
}

#[derive(PartialEq)]
pub struct Residue<P, const L: usize>
where
    P: ResidueParameters<L>,
{
    repr: UInt<L>,
    phantom: PhantomData<P>,
}
```

It could be wrapped up in a macro which defines a ZST and impls the trait for it like so to reduce the boilerplate:

```rust
impl_modulus!(Modulus, "ffffffff00000001...");
```

tarcieri commented:
Also thanks for submitting this! I'll try to review it in depth when I have some time.

jellevos commented:
> It could be wrapped up in a macro which defines a ZST and impls the trait for it like so to reduce the boilerplate:
> 
> ```rust
> impl_modulus!(Modulus, "ffffffff00000001...");
> ```

This is a really good idea! One thing that is not immediately clear to me is how ergonomic this will be for non-hardcoded moduli, but I will investigate.

> Also thanks for submitting this! I'll try to review it in depth when I have some time.

Feel free to wait a bit until I have incorporated the changes above. :) If you like, I can ping you when they are ready.

tarcieri commented:
> This is a really good idea! One thing that is not immediately clear to me is how ergonomic this will be for non-hardcoded moduli, but I will investigate.

The philosophy of this crate is to largely focus on compile-time optimizations, and perhaps do a secondary pass for things that are dependent on runtime values (e.g. #58).

Having a `const MODULUS` unlocks some potential interesting future optimizations when used in conjunction with other `const fn` methods of `UInt`.

It would enable checking special properties of the modulus, and using those properties to select particular implementations in a way this could all be known at compile time. This would allow the compiler to optimize away the branches, so you could provide several implementations of e.g. modular sqrt, picking the best one for a particular modulus at compile time.

jellevos commented:
@tarcieri The implementation is mostly finished, but I am running into a seemingly fundamental issue. Specifically, I get the following error when running the tests in `uint/modular/pow.rs` on stable:
```
error: any use of this value will cause an error
  --> src/uint/shr.rs:67:20
   |
67 |                   if i < (LIMBS - 1) - full_shifts {
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      exceeded interpreter step limit (see `#[const_eval_limit]`)
   |                      inside `uint::shr::<impl uint::UInt<16_usize>>::shr_vartime` at src/uint/shr.rs:67:20
   |                      inside `uint::shr::<impl uint::UInt<16_usize>>::shr_vartime_wide` at src/uint/shr.rs:89:17
   |                      inside `uint::div::<impl uint::UInt<16_usize>>::ct_reduce_wide` at src/uint/div.rs:100:17
   |                      inside `<modular::pow::tests::Modulus as ResidueParams<{ nlimbs!(<$uint_type>::BIT_SIZE) }>>::R2` at src/uint/modular/macros.rs:24:17
   |
  ::: src/uint/modular/macros.rs:23:13
   |
23 | /             const R2: $crate::UInt<{ nlimbs!(<$uint_type>::BIT_SIZE) }> =
24 | |                 $crate::UInt::ct_reduce_wide(Self::R.square_wide(), &Self::MODULUS).0;
   | |______________________________________________________________________________________-
   |
   = note: `#[deny(const_err)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>

error[E0080]: evaluation of `uint::modular::Residue::<uint::modular::pow::tests::Modulus, 16_usize>::new` failed
  --> src/uint/modular/mod.rs:48:41
   |
48 |         let product = integer.mul_wide(&MOD::R2);
   |                                         ^^^^^^^ referenced constant has errors

note: the above error was encountered while instantiating `fn uint::modular::Residue::<uint::modular::pow::tests::Modulus, 16_usize>::new`
  --> src/uint/modular/macros.rs:36:9
   |
36 |         $crate::uint::modular::Residue::<$modulus, { $modulus::LIMBS }>::new($variable)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
  ::: src/uint/modular/pow.rs:56:24
   |
56 |         let base_mod = residue!(base, Modulus);
   |                        ----------------------- in this macro invocation
   |
   = note: this note originates in the macro `residue` (in Nightly builds, run with -Z macro-backtrace for more info)
```

In other words, we are hitting the compiler's `const_eval_limit`, and I do not immediately see a way to fix this. ☹️ On nightly, the fix is simple: I have added the following snippet in `lib.rs`, which basically turns off this check (we could also just put in a larger number than the default, rather than completely omitting the check):
```
#![feature(const_eval_limit)]
#![const_eval_limit = "0"]
```

What do you think? Do you have any clever ideas on how to get around this? Apart from this, I only need to add a few docstrings before the PR should be ready for review.

jellevos commented:
I suppose we could also shrink the modulus in those tests, which would make the tests and library work on stable but would force dependencies that use somewhat large moduli to be nightly.

haslersn commented:
> the workaround that @haslersn suggested which uses a ZST + trait with an associated constant seems OK as a workaround

@tarcieri In practice this was very cumbersome, because one has to drag the `const L` parameter through everything. So instead I ended up creating a trait `GenericUInt` that abstracts over `UInt<L>` and using it like this:

```rust
pub trait ResidueParameters: PartialEq + Debug + Send + 'static {
    type UInt: GenericUInt;
    const MODULUS: Option<Self::UInt>;
}

#[derive(PartialEq)]
pub struct Residue<P>
where
    P: ResidueParameters,
{
    repr: P::UInt,
    phantom: PhantomData<P>,
}
```

tarcieri commented:
@jellevos wow, that is really unfortunate. I was unaware `const_eval_limit` was set so low as to preclude these sorts of computations from working.

I see there's a tracking issue for the low limit precluding other real-world usages: https://github.com/rust-lang/rust/issues/93481 (The main tracking issue for `const_eval_limit` is https://github.com/rust-lang/rust/issues/67217).

Tough to say what to do about this. It would be good to determine that the caller can change this limit, or otherwise if `crypto-bigint` needs to set it, we'd need to add something like a `nightly` crate feature which bumps the limit.

tarcieri commented:
> So instead I ended up creating a trait `GenericUInt` that abstracts over `UInt<L>`

@haslersn FWIW the existing `Integer` trait largely provides that.

haslersn commented:
@tarcieri yes, but the `Integer` trait lacks access to modular arithmetic (`wrapping_add`, ...), direct access to limbs, and conversion from primitive types such as `u32`.

tarcieri commented:
@haslersn you can add those bounds to the existing `Integer` trait, rather than making a new one

haslersn commented:
@tarcieri yes, when adding the functionality directly in crypto-bigint, that would be the way to go.

jellevos commented:
> Tough to say what to do about this. It would be good to determine that the caller can change this limit, or otherwise if `crypto-bigint` needs to set it, we'd need to add something like a `nightly` crate feature which bumps the limit.

I think the best solution in the mean time would be to make sure the tests in this library pass on stable and then leave the problem with the user (i.e. a user must switch to nightly if the const computations exceed the current limit).

I can finish these changes tomorrow, then I would say this is ready for review. I will add a clear statement in the docs that one may need to up `const_eval_limit`.

tarcieri commented:
@jellevos sounds good!

FWIW I left a comment here: https://github.com/rust-lang/rust/issues/93481#issuecomment-1279997537

jellevos commented:
> > the workaround that @haslersn suggested which uses a ZST + trait with an associated constant seems OK as a workaround
> 
> @tarcieri The initial attempt at this was very cumbersome, because one has to drag the `const L` parameter through everything.

Currently, I use a second macro `residue!` that at least prevents the end user from being exposed to this (i.e. having to set `const LIMBS`). It takes a `UInt` and such a modulus trait and returns a `Residue`.

jellevos commented:
@tarcieri Ready for review 😄 I did have to bump the minimum rustc version to 1.61, as the previous versions did not stabilize trait bounds on const functions.

tarcieri commented:
@jellevos just did a quick scan and it looks good at first glance.

I was hoping to do a more in-depth review this weekend but didn't have time, and also you were still making changes.

I'll try to take a look at this soon, hopefully next weekend.

tarcieri commented:
I was hoping to get around to look at this over the weekend but was too busy. Will look at it soon.

jellevos commented:
> The philosophy of this crate is to largely focus on compile-time optimizations, and perhaps do a secondary pass for things that are dependent on runtime values (e.g. #58).
> 
> Having a `const MODULUS` unlocks some potential interesting future optimizations when used in conjunction with other `const fn` methods of `UInt`.
> 
> It would enable checking special properties of the modulus, and using those properties to select particular implementations in a way this could all be known at compile time. This would allow the compiler to optimize away the branches, so you could provide several implementations of e.g. modular sqrt, picking the best one for a particular modulus at compile time.

I did some thinking about this, and from my perspective, it would seem that supporting variable moduli would not directly contradict this crate's philosophy. Those heap-backed integers to me are a different beast: They involve types for which we do not know their concrete size at compile time, but for a variable modulus we can generally assume to know. While the current const moduli enable implementing (a compile-time optimized!) version of e.g. integer-based ElGamal, it would not allow us to implement e.g. RSA or Paillier without knowing the public key ahead of time.

For the crate that I maintain (scicrypt, definitely still WIP) supporting variable moduli would allow me to completely switch away from other crates/libraries and use `crypto-bigint` to implement the cryptosystems, so they are constant-time. I also believe it might be useful for the RSA crate, which then may be able to move away from the not necessarily constant-time `num-bigint` crate.

If you are okay with exploring this direction, I am happy to implement this in one of the following weeks. The API would require some thinking (safety vs ergonomics).

tarcieri commented:
@jellevos given the limitations we've encountered, backing things out to runtime-based moduli seems fine

jellevos commented:
@tarcieri Just for completeness: I only hit the const limit when testing with 2048-bit moduli so the code as it is (without runtime-based moduli) should already be useful for applications with smaller moduli. 😄 Considering this, I will address runtime-based moduli in a second PR if you agree.

tarcieri commented on src/uint/inv_mod.rs:
```diff
@@ -25,11 +27,72 @@ impl<const LIMBS: usize> UInt<LIMBS> {
     }
+
+    /// Computes the multiplicative inverse of `self` mod `modulus`. In other words `self^-1 mod modulus`. Returns `(inverse, 1...1)` if an inverse exists, otherwise `(undefined, 0...0)`. The algorithm is the same as in GMP 6.2.1's `mpn_sec_invert`.
+    pub const fn ct_inv_mod(self, modulus: &UInt<LIMBS>) -> (Self, Word) {
```
This crate has a "constant time by default unless explicitly marked otherwise" policy, so the `ct_*` prefixes on the functions are redundant.

We explicitly mark variable-time functions as `*_vartime`.

I think there were a handful of `pub(crate)` functions with `ct_*` names before which should probably be renamed, but this PR is greatly expanding that and exposing them as part of the public API.

```suggestion
    pub const fn inv_mod(self, modulus: &UInt<LIMBS>) -> (Self, Word) {
```

tarcieri replied:
Oh, scrolling down I now see there's a separate definition of `inv_mod`.

Not sure how better to name the const versions other than `const_*`, but having one named `ct_*` and the other not would generally imply to me one is constant time and the other is not, which is not the case here.

jellevos replied:
My mistake, this function should have been `pub(crate)`. Note that this function is in fact (supposed to be) constant-time. I took the `ct_*` prefix to mean that these functions cannot work with `subtle` because they are `const`, but I see that this is indeed not always consistent. Perhaps this naming issue could be addressed in a future PR?

tarcieri commented on src/uint/inv_mod.rs:
```diff
@@ -25,11 +27,72 @@ impl<const LIMBS: usize> UInt<LIMBS> {
+        let mut u = UInt::ONE;
+        let mut v = UInt::ZERO;
+
+        let mut b = *modulus;
```
If you're derefing anyway, perhaps consider making `modulus` owned instead of borrowed?

tarcieri commented on src/uint/modular/macros.rs:
```diff
@@ -0,0 +1,39 @@
+                .0
+                .wrapping_add(&$crate::UInt::ONE);
+            const R2: $crate::UInt<{ nlimbs!(<$uint_type>::BIT_SIZE) }> =
+                $crate::UInt::ct_reduce_wide(Self::R.square_wide(), &Self::MODULUS).0;
```
If this is the only usage of `square_wide`, it seems like it can be trivially replaced with `Self::R.mul_wide(&Self::R)`

jellevos replied:
Is it okay if I leave it like this so I can implement a faster square method later today? Modular exponentiation would benefit greatly from an optimized squaring method. 😄

tarcieri commented on src/uint/pow_mod.rs:
```diff
@@ -0,0 +1,42 @@
+// use crate::{Concat, Split, UInt};
```
Perhaps remove this file and PR it separately?

jellevos replied:
Oops, my bad. This is old code that was not supposed to be pushed. 😅

jellevos commented:
@tarcieri Ready for review! Later today I expect to finish the efficient squaring method in a separate PR.

jellevos commented:
**Summary of new changes:**
- Changed folder structure to accommodate runtime-based moduli neatly.
- I forgot to implement the inverse for residues before; that's done now.
- Renamed `Residue` to `ConstResidue`, I reserved `Residue` for residues with a modulus decided at runtime.
- Added traits to provide a consistent API between `ConstResidue`s and `Residue`s.

I want to raise an issue about the multiplicative inverses as they are currently very unsafe to use. Currently, I only check that an inverse exists using `debug_assert!`. Release builds, however, do not panic when an inverse does not exist. Instead, they actually continue on with computations on a residue with an undefined value.

@tarcieri Do you have any clever ideas on how to handle this? One idea is to return a `CtOption`, but then the inverse cannot be a `const fn`. Thanks for your help!

tarcieri commented:
> Renamed `Residue` to `ConstResidue`, I reserved `Residue` for residues with a modulus decided at runtime.

I'd personally prefer (and it would be more consistent with the other RustCrypto projects) if it were the reverse of that, with `Residue` being const-friendly, and something like `DynResidue` with a runtime-dependent modulus.

jellevos commented:
I agree, I reverted it. 😄 I am still thinking about how to best handle the problem with multiplicative inverses:

> I want to raise an issue about the multiplicative inverses as they are currently very unsafe to use. Currently, I only check that an inverse exists using debug_assert!. Release builds, however, do not panic when an inverse does not exist. Instead, they actually continue on with computations on a residue with an undefined value.
>
> @tarcieri Do you have any clever ideas on how to handle this? One idea is to return a CtOption, but then the inverse cannot be a const fn. Thanks for your help!

tarcieri commented:
@jellevos yeah, that's a tricky and annoying tradeoff.

You could have a `const fn` version that panics, and a non-`const fn` version that returns a `CtOption`, which are both wrappers around a third internal (`const fn` version that does something like return the `Choice` selector as a `u8`, but that's messy.

tarcieri commented:
Cool, will take another look this weekend

haslersn commented:
When using this much appreciated feature in my project, I needed to do some changes to the API surface around `Residue`. Some changes were necessary to make it usable at all (from different crates) and some of them were meant to simplify usage of Residues, especially in generic contexts.

I'd like to PR these changes, discuss them, and reach a consensus. However, I'll only get to that after the paper deadline, which is in about a week.

So it would be nice if you could wait a bit with the next release.

tarcieri commented:
No worries there, it's going to be at least a few months before we're ready to cut another release
